Quests/Kanto/CascadeBadgeQuest.lua

```diff
@@ -26,28 +30,39 @@ function CascadeBadgeQuest:new()
 end
 
 function CascadeBadgeQuest:isDoable()
-	if not hasItem("Cascade Badge") and self:hasMap()
-	then
+	if self:hasMap() then
 		return true
 	end
 	return false
 end
 
 function CascadeBadgeQuest:isDone()
-	if hasItem("Cascade Badge") then
+	if getMapName() == "Route 5" then
 		return true
 	else
 		return false
 	end
 end
 
-function CascadeBadgeQuest:CeruleanCity()
+function CascadeBadgeQuest:CeruleanCity()	
 	if self:needPokecenter() or not game.isTeamFullyHealed() or self.registeredPokecenter ~= "Pokecenter Cerulean" then
 		return moveToMap("Pokecenter Cerulean")
 	elseif not self:isTrainingOver() then
 		return moveToCell(39,0)-- Route 24 Bridge
 	else
-		return moveToMap("Cerulean Gym")
+		if not hasItem("Cascade Badge") then
+			return moveToMap("Cerulean Gym")
+		else  -- Gym complete --> Get Ticket Bill
```

Comment:
This should be done before the arena.  
We have to ensure it is impossible for the bot to finish the quest without the ticket. Any time this quest start, you should check if Bill already gave us the ticket.  
That way a crash won't skip the ticket.